leftist.ml

```diff
@@ -0,0 +1,71 @@
+(*  Autor kodu: Przemysław Podleśny *)
+(*  Code reviewer: Jakub Nowak      *)
+
+(*  Korzeń kolejki priorytetowej - albo drzewo puste,
+    albo rekord reprezentujący węzeł drzewa;
+    left - lewy syn, priority - priorytet, right - prawy syn,
+    height - wysokość drzewa                                    *)
+type 'a queue =
+    | EmptyQueue
+    | Node of {
+        left: 'a queue;
+        priority: 'a;
+        right: 'a queue;
+        height: int;
+    }
+
+(*  Zwraca prawą wysokość wierzchołka "q"
+    (0 dla liścia, -1 dla pustego).         *)
+let height = function
+    | EmptyQueue -> -1
+    | Node {height} -> height
+
+(*  Tworzy drzewo o lewym synu w l,
+    priorytecie priority i prawym synu w r. *)
+let make_node l priority r =
+    Node {
+        left = l;
+        priority = priority;
+        right = r;
+        height = height r + 1;
+    }
+
+(*  Porządkuje kolejki q1, q2 nierosnąco
+    pod względem prawej wysokości ich korzeni.  *)
+let order_heights q1 q2 =
+    if height q1 >= height q2 then q1, q2
+    else q2, q1
+
+(*  Pusta kolejka.  *)
+let empty = EmptyQueue
+
+(*  Sprawdzenie, czy kolejka jest pusta.    *)
+let is_empty = function
+    | EmptyQueue -> true
+    | Node _ -> false
+
+(*  Łączenie kolejek q1 i q2.   *)
+let rec join q1 q2 =
+    match q1, q2 with (* Gdy jedna jest pusta, nie ma co łączyć.    *)
+    | EmptyQueue, _ -> q2
+    | _, EmptyQueue -> q1
+    | Node {left; priority = prio1; right}, Node {priority = prio2} ->
+        if prio1 > prio2 then join q2 q1 (* Niech prio1 < prio2. *)
```

Comment:
Według mnie ładniej by było coś w ten deseń:
```ocaml
if prio1 > prio2 then join_helper q2 q1 else join_helper q1 q2
```
Aczkolwiek to moje subiektywne odczucie.

Reply:
Szczerze mówiąc, to mnie też się takie rozwiązanie średnio podobało. Zaraz spróbuję to zmienić.

Reply:
Ok, łatwiej by było z krotkami, a tak to tylko by się za bardzo wydłużyło. Za to teraz jest na początku funkcji, więc zbędne matchowanie będzie wykonane co najwyżej raz (optymalizacja).

Reply:
Optymalizacja jest mało istotna. Dobrą zasadą jest, że pisząć kod należy się skupić na czytelności i nie przejmować się drobnymi optymalizacjami. Ewentualnie optymalizacje robi się później, po zindetyfikowaniu, który kod powoduje wolne działanie aplikacji i tej optymalizacji wymaga.

Reply:
Lepiej chyba wrzucić te dwie linijki z `EmptyQueue` na początek `match`. Moim zdaniem tak będzie ładniej.